src/kernel.cc

```diff
@@ -741,33 +783,101 @@ void quantum_kernel::gate(const Str &gname, UInt q0, UInt q1) {
 }
 
 /**
- * custom gate with arbitrary number of operands
- * check qubit/creg indices against platform parameters; fail fatally if an index is out of range
- * find matching gate in kernel's and platform's gate_definition; when no match, fail
+ * general user-level gate creation with any combination of operands
+ *
+ * check argument register indices against platform parameters; fail fatally if an index is out of range
+ * add implicit arguments if absent (used when no register argument means all registers)
+ * find matching gate in kernel's and platform's gate_definition (custom or default); when no match, fail
  * return the gate (or its decomposition) by appending it to kernel.c, the current kernel's circuit
  */
 void quantum_kernel::gate(
     const Str &gname,
     const Vec<UInt> &qubits,
     const Vec<UInt> &cregs,
     UInt duration,
-    Real angle
+    Real angle,
+    const Vec<UInt> &bregs,
+    cond_type_t gcond,
+    const Vec<UInt> &gcondregs
 ) {
-    /// @todo-rn: move these check to a platform-specific backend after qubits are initialized
+    QL_DOUT("gate:" <<" gname=" << gname <<" qubits=" << qubits <<" cregs=" << cregs <<" duration=" << duration <<" angle=" << angle <<" bregs=" << bregs <<" gcond=" << gcond <<" gcondregs=" << gcondregs);
+
     for (auto &qno : qubits) {
         if (qno >= qubit_count) {
             QL_FATAL("Number of qubits in platform: " << to_string(qubit_count) << ", specified qubit numbers out of range for gate: '" << gname << "' with qubits " << qubits);
         }
     }
-
-    for (auto & cno : cregs) {
+    for (auto &cno : cregs) {
         if (cno >= creg_count) {
             QL_FATAL("Out of range operand(s) for '" << gname << "' with cregs " << cregs);
         }
     }
+    for (auto &bno : bregs) {
+        if (bno >= breg_count) {
+            QL_FATAL("Out of range operand(s) for '" << gname << "' with bregs " << bregs);
+        }
+    }
+    if (!gate::is_valid_cond(gcond, gcondregs)) {
+        QL_FATAL("Condition " << gcond << " of '" << gname << "' incompatible with gcondregs " << gcondregs);
+    }
+    for (auto &cbno : gcondregs) {
+        if (cbno >= breg_count) {
+            QL_FATAL("Out of range condition operand(s) for '" << gname << "' with gcondregs " << gcondregs);
+        }
+    }
+    auto lqubits = qubits;
+    auto lcregs = cregs;
+    auto lbregs = bregs;
+    gate_add_implicits(gname, lqubits, lcregs, duration, angle, lbregs, gcond, gcondregs);
+    if (!gate_nonfatal(gname, lqubits, lcregs, duration, angle, lbregs, gcond, gcondregs)) {
+        QL_FATAL("Unknown gate '" << gname << "' with qubits " << lqubits);
+    }
+}
+
+/**
+ * preset condition to make all future created gates conditional gates with this condition
```

Comment:
This looks like a hack to me, wouldn't it be better to add the support for conditional gates where needed

Reply:
It is meant for gates that are decomposed by passes instead of steered by the config file's decompositions,
e.g. toffoli_decomposition and k.clifford(). I didn't want to clutter up those codes with conditional execution
but instead to do it under the hood.
We could argue about having it as a Python interface.